jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java

```diff
@@ -779,22 +779,25 @@ public static String getCanonicalNameForWebAppTmpDir(WebAppContext context)
                 resource = context.newResource(context.getWar());
             }
 
-            String tmp = URIUtil.decodePath(resource.getURI().getPath());
+            String tmp = getResourceBasePath(resource);
```

Comment:
This shouldn't decode (yet) here.

The check for last "/' and "!" need to occur on the raw (non-decoded) URI.